channels/rdpdr/server/rdpdr_main.c

```diff
@@ -591,37 +825,968 @@ static int rdpdr_server_stop(RdpdrServerContext* context)
+    Stream_Write_UINT32(s, 3); /* SharedAccess (4 bytes) */
+    Stream_Write_UINT32(s, createDisposition); /* CreateDisposition (4 bytes) */
+    Stream_Write_UINT32(s, createOptions); /* CreateOptions (4 bytes) */
+    Stream_Write_UINT32(s, pathLength); /* PathLength (4 bytes) */
+
+	/* Convert the path to Unicode. */
+	MultiByteToWideChar(CP_ACP, 0, path, -1, (LPWSTR) Stream_Pointer(s), pathLength);
+	Stream_Seek(s, pathLength);
+
+    Stream_SealLength(s);
+
+	status = WTSVirtualChannelWrite(context->priv->ChannelHandle, (PCHAR) Stream_Buffer(s), Stream_Length(s), &written);
+
+    Stream_Free(s, TRUE);
+
+    return status;
+}
+
+static BOOL rdpdr_server_send_device_close_request(
+	RdpdrServerContext* context,
+	UINT32 deviceId,
+	UINT32 fileId,
+	UINT32 completionId
+)
+{
+	ULONG written;
+	BOOL status;
+	wStream* s;
+
+	WLog_DBG(TAG, "%s: deviceId=%d, fileId=%d", __FUNCTION__, deviceId, fileId);
+
+	s = Stream_New(NULL, 128);
+
+	rdpdr_server_write_device_iorequest(s, deviceId, fileId, completionId, IRP_MJ_CLOSE, 0);
+
+	Stream_Zero(s, 32); /* Padding (32 bytes) */
+
+    Stream_SealLength(s);
+
+	status = WTSVirtualChannelWrite(context->priv->ChannelHandle, (PCHAR) Stream_Buffer(s), Stream_Length(s), &written);
+
+    Stream_Free(s, TRUE);
+
+    return status;
+}
+
+static BOOL rdpdr_server_send_device_read_request(
+	RdpdrServerContext* context,
+	UINT32 deviceId,
+	UINT32 fileId,
+	UINT32 completionId,
+	UINT32 length,
+	UINT32 offset
+)
+{
+	ULONG written;
+	BOOL status;
+	wStream* s;
+
+	WLog_DBG(TAG, "%s: deviceId=%d, fileId=%d, length=%d, offset=%d", __FUNCTION__, deviceId, fileId, length, offset);
+
+	s = Stream_New(NULL, 128);
+
+	rdpdr_server_write_device_iorequest(s, deviceId, fileId, completionId, IRP_MJ_READ, 0);
+
+	Stream_Write_UINT32(s, length); /* Length (4 bytes) */
+	Stream_Write_UINT32(s, offset); /* Offset (8 bytes) */
+	Stream_Write_UINT32(s, 0);
+	Stream_Zero(s, 20); /* Padding (20 bytes) */
+
+    Stream_SealLength(s);
+
+	status = WTSVirtualChannelWrite(context->priv->ChannelHandle, (PCHAR) Stream_Buffer(s), Stream_Length(s), &written);
+
+    Stream_Free(s, TRUE);
+
+    return status;
 }
+
+static BOOL rdpdr_server_send_device_write_request(
+	RdpdrServerContext* context,
+	UINT32 deviceId,
+	UINT32 fileId,
+	UINT32 completionId,
+	const char* data,
+	UINT32 length,
+	UINT32 offset
+)
+{
+	ULONG written;
+	BOOL status;
+	wStream* s;
+
+	WLog_DBG(TAG, "%s: deviceId=%d, fileId=%d, length=%d, offset=%d", __FUNCTION__, deviceId, fileId, length, offset);
+
+	s = Stream_New(NULL, 64 + length);
+
+	rdpdr_server_write_device_iorequest(s, deviceId, fileId, completionId, IRP_MJ_WRITE, 0);
+
+	Stream_Write_UINT32(s, length); /* Length (4 bytes) */
+	Stream_Write_UINT32(s, offset); /* Offset (8 bytes) */
+	Stream_Write_UINT32(s, 0);
+	Stream_Zero(s, 20); /* Padding (20 bytes) */
+	Stream_Write(s, data, length); /* WriteData (variable) */
+
+    Stream_SealLength(s);
+
```

Comment:
Looks like a formatting problem. - Some lines are more intended on the same code level than others.